src/Microsoft.Data.Analysis/DataFrame.Join.cs

```diff
@@ -293,8 +297,14 @@ public DataFrame Merge<TKey>(DataFrame other, string leftJoinColumn, string righ
                     {
                         foreach (long supplementaryRowIndex in rowIndices)
                         {
-                            retainedRowIndices.Append(i);
-                            supplementaryRowIndices.Append(supplementaryRowIndex);
+                            long? leftDataFrameIndex = isLeftDataFrameRetained ? i : supplementaryRowIndex;
+                            long? rightDataFrameIndex = isLeftDataFrameRetained ? supplementaryRowIndex : i;
```

Comment:
nit: This is inside a loop, should we try to optimize this a little by checking once? 

```C#
(long? leftDataFrameIndex, long? rightDataFrameIndex) = isLeftDataFrameRetained ? (i, supplementaryRowIndex) : (supplementaryRowIndex, i)?
```